### ISSUE TITLE
The Check function returns exit(-1) which exits the Python interpreter

### ISSUE THREAD

**Issue by jseabold** [dmlc/xgboost#356]
```
In [1]: y = np.random.randint(2, size=100)

In [2]: X = np.random.randn(100, 10)

In [3]: from xgboost import XGBClassifier

In [4]: mod = XGBClassifier(colsample_bytree=.05)

In [5]: mod.fit(X, y)
colsample_bytree=0.05 is too small that no feature can be included
```

And exits the interpreter. This happens in other places whenever Check return `exit(-1)`.

**Comment by tqchen** [comment]
this can be fixed. If there is a c function from python lib that allows us
to throw exception from shared lib. Actually we have a custom error msg
that allows redirection of error reporting. So in R it throws an error
instead. The only problem was I am not very familiar whether there exists
such hook in python

**Comment by jseabold** [comment]
Is that the stuff in `R-package/src/xgboost_assert.c`?

**Comment by tqchen** [comment]
There is a error function in R runtime lib that can be called todo the
things

**Comment by tqchen** [comment]
OK, here is one solution that can solve this problem, I am writing it down in case we want to do it in future. Basically, it seems to be hard to pass exception out from ctypes.

One possible solution: Rewrap the current python interface using cython, this means take the functions in wrapper.h and do the wrap. There wrapper need to be recompiled into a C++ module(by removing extern C with a macro switch), and customized the behavior of Check to throw an exception instead, cython will allow the exception to pass through and translated it into a python exception.

This do require a bit work to do and I do not have time recently on this. So I am putting the solution down here in case anyone is interested in take a stab on this

Tianqi

**Comment by tqchen** [comment]
Another way would be to standardize current C interface, make every function return a integer referring to whether the result is successful or not. Catch the exception in the wrapper function and handle it properly, this allows us to keep the current ctypes interface without introducing cython dependency

**Comment by tqchen** [comment]
should be fixed by #373
